_drafts/ota-delta-updates.md

```diff
@@ -0,0 +1,600 @@
+implementation from top to bottom.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+## Delta Updates Explained
+
+> Note: the next few sections provide an overview of delta updates, why it
+> matters, and how it works. If you'd like to skip straight to implementation,
+> click
+> [here](#delta-updates-implementation).
+
+The concept behind delta updates is intuitive. When we put together a new
+version of our firmware, we do not change every single line of code. Instead, we
+edit specific sections of the code to change the behavior of the code and
+introduce a new feature or fix a bug.
+
+This means that our new firmware is largely made up of the old firmware, plus a
+few changes.
+
+<img width=600px src="{% img_url fwup-delta/new-code-old-image.png %}" />
+
+Wouldn't it be neat if we could collect only the code changes and send those
+down to the device, rather than the full firmware update? This is exactly what
+delta updates allow us to do.
+
+The process is broken up into two steps. First, we need to extract the bits that
+have changed from our new image, we'll call it DIFF.
+
+<img width=600px src="{% img_url fwup-delta/new-code-DIFF.png %}" />
+
+Then we reconstruct our image from our original firmware and our delta image,
+which we'll call PATCH.
+
+<img width=700px src="{% img_url fwup-delta/new-code-PATCH.png %}" />
+
+### Pros and Cons of Delta Updates
+
+The obvious advantage of delta updates is the small size of the resulting image.
+Delta images are often one to two orders of magnitude smaller than full system
+images. The reduction in size has multiple beneficial effects:
+
+1. OTA updates become possible over very low bandwidth links (e.g. LoRaWAN).
+2. Updates are much faster over low bandwidth links. For example, a 10MB image
+   may take over 15 minutes to download over a BLE connection to a mobile phone,
+   even at [peak throughput]({% post_url 2019-09-24-ble-throughput-primer %}). A
+   delta update would take less than 1 minute to download, leading to a much
+   better customer experience and less risk of power loss mid-update.
+3. The lifetime of flash memory can be extended, as fewer writes are needed to
+   install a delta image than a full image.
+4. OTA consumer less power, thanks to the reduced communication and flash
+   writing required.
+
+However, there is no such thing as a free lunch! Delta updates come with some
+downsides worth considering:
+
+1. More complexity! Your firmware update code now needs to implement the
+   patching algorithm, which creates more opportunities for bugs.
+2. Delta updates requires more code space to implement the patching algorithm,
+   so if you are constrained already it may not be appropriate for your project.
+3. Managing delta updates creates more complexity on the backend and on the
+   release process.
+
+That last point bears elaborating a bit. Unlike all system images, each delta
```

Comment:
Algorithms like casync, zsync are designed to have the client compute the delta, not sure if it's worth mentioning though

Reply:
Decided not to include this